File: algo_expert/dynamic_programming/ways_to_make_change.py




def numberOfWaysToMakeChange(n, denoms):
    # Write your code here.
    ways = [0 for _ in range(n+1)]
    ways[0] = 1
    for denom in denoms:
        for amount in range(1, n+1):
            if denom <= amount:
                ways[amount] += ways[amount-denom]
    return ways[-1]


denoms = [1, 5, 10]
target = 10
res = numberOfWaysToMakeChange(target, denoms)
print(f"donoms {denoms} target={target}")
print(f"numberOfWaysToMakeChange {res}")

File: algo_expert/dynamic_programming/ways_to_make_change.js


function numberOfWaysToMakeChange(n, denoms) {
    
    let ways = new Array(n+1).fill(0);
    ways[0] = 1;
    for (denom of denoms){
        for (amount=1;amount<n+1;amount++){
            if (denom <= amount){
                ways[amount] += ways[amount-denom];
            }
        }
    }
    return ways[ways.length - 1]
  }




const denoms = [1, 5, 10];
const target = 10;
const res = numberOfWaysToMakeChange(target, denoms);

console.log(`denoms ${denoms} target=${target}`);
console.log(`numberOfWaysToMakeChange ${res}`);